### ISSUE TITLE
gcp_compute plugin does not work

### ISSUE THREAD

**Issue by jesusch** [ansible/ansible#44404]
##### SUMMARY
gcp_compute plugin does not work. always an error that no matching inventory plugin is working

##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
gcp_compute

##### ANSIBLE VERSION
```
ansible 2.6.3
  config file = /Users/jesusch/.ansible.cfg
  configured module search path = [u'/Users/jesusch/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.15 (default, Jul 23 2018, 21:27:06) [GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.2)]
```

##### CONFIGURATION
DEFAULT_HOST_LIST(/Users/jesusch/.ansible.cfg) = [u'/Users/jesusch/ansible/inventory']
DEFAULT_ROLES_PATH(/Users/jesusch/.ansible.cfg) = [u'/Users/jesusch/ansible/roles']
HOST_KEY_CHECKING(/Users/jesusch/.ansible.cfg) = False
RETRY_FILES_ENABLED(/Users/jesusch/.ansible.cfg) = False

##### OS / ENVIRONMENT
Darwin Bjorns-MacBook-Pro.local 17.7.0 Darwin Kernel Version 17.7.0: Thu Jun 21 22:53:14 PDT 2018; root:xnu-4570.71.2~1/RELEASE_X86_64 x86_64

##### STEPS TO REPRODUCE
Configured GCP plugin as described in https://docs.ansible.com/ansible/2.6/scenario_guides/guide_gce.html

```yaml
---
plugin: gcp_compute
projects:
  - google.com:myProject-1337
filters: null
auth_kind: serviceaccount
service_account_file: /Users/jesusch/Downloads/myProject-1337-somehash.json
```

##### EXPECTED RESULTS
ansible-inventory --list -i inventory/myproject.gcp.yml should list my GC hosts and its metadata

##### ACTUAL RESULTS
```
ANSIBLE_DEBUG=1 ansible-inventory --list -i inventory/myproject.gcp.yml
 40804 1534774235.58981: starting run
 40804 1534774235.68904: Added group all to inventory
 40804 1534774235.68913: Added group ungrouped to inventory
 40804 1534774235.68919: Group all now contains ungrouped
 40804 1534774235.70007: Loading InventoryModule 'host_list' from /usr/local/lib/python2.7/site-packages/ansible/plugins/inventory/host_list.py
 40804 1534774235.70106: assigned :doc
 40804 1534774235.70117: assigned :plainexamples
 40804 1534774235.70164: Loading InventoryModule 'script' from /usr/local/lib/python2.7/site-packages/ansible/plugins/inventory/script.py
 40804 1534774235.70398: assigned :doc
 40804 1534774235.70418: Loaded config def from plugin (inventory/script)
 40804 1534774235.70446: Loading InventoryModule 'yaml' from /usr/local/lib/python2.7/site-packages/ansible/plugins/inventory/yaml.py
 40804 1534774235.70576: assigned :doc
 40804 1534774235.70584: assigned :plainexamples
 40804 1534774235.70603: Loaded config def from plugin (inventory/yaml)
 40804 1534774235.70668: Loading InventoryModule 'ini' from /usr/local/lib/python2.7/site-packages/ansible/plugins/inventory/ini.py
 40804 1534774235.70865: assigned :doc
 40804 1534774235.70874: assigned :plainexamples
 40804 1534774235.70911: Loading InventoryModule 'auto' from /usr/local/lib/python2.7/site-packages/ansible/plugins/inventory/auto.py
 40804 1534774235.70965: assigned :doc
 40804 1534774235.70971: assigned :plainexamples
 40804 1534774235.70984: Examining possible inventory source: /Users/jesusch/git/ansible/inventory/myproject.gcp.yml
 40804 1534774235.70992: Attempting to use plugin host_list (/usr/local/lib/python2.7/site-packages/ansible/plugins/inventory/host_list.py)
 40804 1534774235.70998: /Users/jesusch/git/ansible/inventory/myproject.gcp.yml did not meet host_list requirements
 40804 1534774235.71004: Attempting to use plugin script (/usr/local/lib/python2.7/site-packages/ansible/plugins/inventory/script.py)
 40804 1534774235.71764: /Users/jesusch/git/ansible/inventory/myproject.gcp.yml was not parsable by script
 40804 1534774235.71790: Attempting to use plugin yaml (/usr/local/lib/python2.7/site-packages/ansible/plugins/inventory/yaml.py)
 40804 1534774235.72000: Loading data from /Users/jesusch/git/ansible/inventory/myproject.gcp.yml
 40804 1534774235.72153: /Users/jesusch/git/ansible/inventory/myproject.gcp.yml was not parsable by yaml
 40804 1534774235.72162: Attempting to use plugin ini (/usr/local/lib/python2.7/site-packages/ansible/plugins/inventory/ini.py)
 40804 1534774235.72316: set inventory_file for ---
 40804 1534774235.72330: set inventory_dir for ---
 40804 1534774235.72335: Added host --- to inventory
 40804 1534774235.72343: Added host --- to group ungrouped
 40804 1534774235.72386: /Users/jesusch/git/ansible/inventory/myproject.gcp.yml was not parsable by ini
 40804 1534774235.72393: Attempting to use plugin auto (/usr/local/lib/python2.7/site-packages/ansible/plugins/inventory/auto.py)
 40804 1534774235.72421: Loading data from /Users/jesusch/git/ansible/inventory/myproject.gcp.yml
 40804 1534774235.77925: Loading InventoryModule 'gcp_compute' from /usr/local/lib/python2.7/site-packages/ansible/plugins/inventory/gcp_compute.py
 40804 1534774235.79347: assigned :doc
 40804 1534774235.79357: assigned :plainexamples
 40804 1534774235.79887: Loading ModuleDocFragment 'constructed' from /usr/local/lib/python2.7/site-packages/ansible/utils/module_docs_fragments/constructed.py
 40804 1534774235.79947: Loading ModuleDocFragment 'inventory_cache' from /usr/local/lib/python2.7/site-packages/ansible/utils/module_docs_fragments/inventory_cache.py
 40804 1534774235.80017: Loaded config def from plugin (inventory/gcp_compute)
 40804 1534774235.80042: Loading data from /Users/jesusch/git/ansible/inventory/myproject.gcp.yml
 40804 1534774236.20005: auto failed to parse /Users/jesusch/git/ansible/inventory/myproject.gcp.yml
```

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/plugins/inventory/gcp_compute.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/inventory/gcp_compute.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by apettinen** [comment]
I hit this same error, but got it working via the following steps:

- enable the plugin in your ansible.cfg (most inventory plugins are off by default), e.g.:
```
[inventory]
enable_plugins = gcp_compute, host_list, yaml, ini, script
```
see: https://docs.ansible.com/ansible/2.5/plugins/inventory.html for more details
Note, this also specifies the search order.
- The source code [lib/ansible/plugins/inventory/gcp_compute.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/inventory/gcp_compute.py) contains a better example how to configure:
```
plugin: gcp_compute
zones: # populate inventory with instances in these regions
  - us-east1-a
projects:
  - gcp-prod-gke-100
  - gcp-cicd-101
filters:
  - machineType = n1-standard-1
  - scheduling.automaticRestart = true AND machineType = n1-standard-1
scopes:
  - https://www.googleapis.com/auth/compute
service_account_file: /tmp/service_account.json
auth_kind: serviceaccount
```

Also ensure your service account has enough permissions to access the desired resources.

This is on ansible 2.6.3 with python 2.7.10.
So at least the docs should be updated to mention the inventory plugins (or at least link to the docs).

**Comment by jesusch** [comment]
Also explicit specifing the enable_plugins sort order does not do the trick
see: 

```
 61631 1535544664.51556: Considering mycloud.gcp.yml
 61631 1535544664.51567: Examining possible inventory source: /Users/jesusch/ansible/inventory/mycloud.gcp.yml
 61631 1535544664.51578: Attempting to use plugin gcp_compute (/usr/local/lib/python2.7/site-packages/ansible/plugins/inventory/gcp_compute.py)
 61631 1535544664.51610: Loading data from /Users/jesusch/ansible/inventory/mycloud.gcp.yml
 61631 1535544664.78010: gcp_compute failed to parse /Users/jesusch/ansible/inventory/mycloud.gcp.yml
 61631 1535544664.79291: /Users/jesusch/ansible/inventory/mycloud.gcp.yml was not parsable by script
 [WARNING]:  * Failed to parse /Users/jesusch/ansible/inventory/mycloud.gcp.yml with gcp_compute plugin: 'NoneType' object has no attribute '__getitem__'
 61631 1535544664.79513: parsed /Users/jesusch/ansible/inventory/mycloud.gcp.yml as False
```

**Comment by jesusch** [comment]
still an issue with 2.6.4

**Comment by jesusch** [comment]
still an issue with 2.7.0

**Comment by bhundven** [comment]
Issue #45105 is a duplicate of this issue.

I have also tested with 2.7.1, and the issue persists.

**Comment by walterdolce** [comment]
I am also having this problem. Adding the `[inventory]` config section didn't work. Ansible version is `2.7.0`.

**Comment by walterdolce** [comment]
Quick update. For me the issue turned out to be I was trying to use a variable inside the dynamic inventory file. 

This did not work because it was causing an "internal" Ansible error, not exposed even if `ANSIBLE_DEBUG` was set:
```
---
plugin: gcp_compute
# ...
service_account_file: "{{ ansible_env.PWD }}/.gcp/svca-key.json"
auth_kind: serviceaccount
# ...
```

And this wasn't working either, for the same reason:
```
---
plugin: gcp_compute
# ...
service_account_file: "{{ some_group_var_I_defined_containing_the_path_to_the_svc_key }}"
auth_kind: serviceaccount
# ...
```

@bcoca @bhundven  Is it expected that Ansible does not compile/expand/parse variables defined in the GCP dynamic inventory and/or "built-in" variables?

**Comment by bhundven** [comment]
So in your `ansible.cfg` you need to enable the gcp_compute plugin:

```
[inventory]
# enable inventory plugins, default: 'host_list', 'script', 'yaml', 'ini'
#enable_plugins = host_list, virtualbox, yaml, constructed
enable_plugins = host_list, script, yaml, ini, gcp_compute
```

This fixed this plugin for me.

**Comment by jesusch** [comment]
also does not work for me:
```
  2896 1541067684.30548: gcp_compute failed to parse /Users/jesusch/ansible/inventory/wywy.gcp.yml
 [WARNING]:  * Failed to parse /Users/jesusch/ansible/inventory/wywy.gcp.yml with ini plugin: /Users/jesusch/ansible/inventory/wywy.gcp.yml:2: Expected key=value host variable assignment, got:
gcp_compute

 [WARNING]:  * Failed to parse /Users/jesusch/ansible/inventory/wywy.gcp.yml with yaml plugin: Plugin configuration YAML file, not YAML inventory

 [WARNING]:  * Failed to parse /Users/jesusch/ansible/inventory/wywy.gcp.yml with constructed plugin: Incorrect plugin name in file: gcp_compute

 [WARNING]:  * Failed to parse /Users/jesusch/ansible/inventory/wywy.gcp.yml with gcp_compute plugin: 'NoneType' object has no attribute '__getitem__'

 [WARNING]: Unable to parse /Users/jesusch/ansible/inventory/wywy.gcp.yml as an inventory source

  2896 1541067684.30767: parsed /Users/jesusch/ansible/inventory/wywy.gcp.yml as False
```

**Comment by chad-autry** [comment]
The example config on the ansible page is misleading or incorrect. Remove "google.com:" just need the project identifier

```
---
plugin: gcp_compute
projects:
  - myProject-1337
...
```

**Comment by jesusch** [comment]
@chad-autry this finally did the trick

documentation update should do the trick

**Comment by Akasurde** [comment]
@jesusch @chad-autry Can you confirm if #50183 works for you ? Thanks.

needs_info

**Comment by jesusch** [comment]
@Akasurde works for me :)

**Comment by sudomann** [comment]
Such conflicting documentation;
[docs here](https://docs.ansible.com/ansible/latest/scenario_guides/guide_gce.html#gce-dynamic-inventory) give the wrong format for project names:
```yaml
projects:
  - google.com:graphite-playground
``` 
I got it to work with the following (notice no `google.com:`):
```yaml
plugin: gcp_compute
projects:
  - my-gcp-project-id
filters:
scopes:
  - https://www.googleapis.com/auth/compute
service_account_file: /path/to/ansible-inventory-plugin-service-account.json
auth_kind: serviceaccount
```
This was based off other replies I saw on issue related to this, and the [docs here](https://docs.ansible.com/ansible/latest/plugins/inventory/gcp_compute.html) which has it properly labelled.

**Comment by Akasurde** [comment]
@sudomann PR https://github.com/ansible/ansible/pull/50183 is yet to merge. That is why you are seeing this conflicts in documentation. Once PR is merged, this conflict will go away.

**Comment by jeremyarr** [comment]
also another gotcha: the file must have a suffix gcp.yml

**Comment by Akasurde** [comment]
@jeremyarr This is already mentioned in documentation - https://docs.ansible.com/ansible/latest/plugins/inventory/gcp_compute.html#synopsis
